### ISSUE TITLE
DistributedDataParallel warning in case of different models on different nodes

### ISSUE THREAD

**Issue by kefirski** [pytorch/pytorch#32426]
## 🚀 Feature
Adding a warning/error message if you had created DDP module from different models.

## Motivation
Recently I faced weird behavior of DDP module – in case if you have different models (i.e., some weights differed) on different nodes, you will receive no error/warning message about this while the whole training process is going to halt.

I forgot that some pre-trained models differed in shapes (the first node had an embedding layer with shape [100, 500], and the second one had an embedding layer with shape [10000, 500]). I had received no error message while my model halted during the training before first forward propagation. 

Thus I had to debug this case for about a day, while it could be significantly easier if I had received a warning message that my models differs.

## Pitch

I'm not sure what is the reason to allow somebody to create DDP model from different modules. I believe that users should receive a warning about that.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski

**Comment by mrshenli** [comment]
> Recently I faced weird behavior of DDP module – in case if you have different models (i.e., some weights differed) on different nodes, you will receive no error/warning message about this while the whole training process is going to halt.

This is true. `DistributedDataParallel` currently does not check model equivalence at construction time, and go ahead and launch `allreduce` on gradient buckets. If that number/size of buckets does not match, it could hang.  

> I'm not sure what is the reason to allow somebody to create DDP model from different modules. I believe that users should receive a warning about that.

Good point! The down side is that doing this check (e.g., through gather/allgather) would introduce extra overhead, and I am a little reluctant to introduce another flag for it. @pietern @zhaojuanmao any suggestions on what's the best way to introduce this check?

**Comment by zhaojuanmao** [comment]
@mrshenli we can extend the debug mode proposal in #22071 to debug these issues

**Comment by mrshenli** [comment]
@zhaojuanmao that indeed sounds better than introducing one more flag.

**Comment by kefirski** [comment]
@mrshenli Adding checks for consistency of the model between several processes also seems to be a logical solution for me.

Furthermore, since this check would be performed only in debug mode, it is not going to hurt the performance in the general case.

**Comment by rohan-varma** [comment]
Self-assigning because it is related to the validation that would be useful for https://github.com/pytorch/pytorch/issues/43690

**Comment by rohan-varma** [comment]
This was added as part of TORCH_DISTRIBUTED_DEBUG, closing it out.
